### PR TITLE
Implement ngx.shared.DICT.decr method

### PR DESCRIPTION
**Pull request by p0pr0ck5**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

Implement a counterpart to ngx.shared.DICT.incr that decrements
the numeric value at a given key, with an optional init. In addition,
decr() cannot receive a negative value, cannot decrement past 0,
and cannot initialize a value less than zero.

See #826
for more details and initial proposal.

### CODE REVIEW

File: t/043-shdict.t [hunk 1]
```diff
@@ -2448,3 +2448,216 @@ GET /test
...
+--- http_config
+    lua_shared_dict dogs 1m;
+--- config
+    location = /test {
+        content_by_lua '
+            local dogs = ngx.shared.dogs
+            dogs:set("bar", 3, 0.001)
+            dogs:set("baz", 2, 0.001)
+            dogs:set("foo", 32, 0.001)
+            ngx.location.capture("/sleep/0.002")
```
Comment (membphis) [inline t/043-shdict.t]: How about use `ngx.sleep` directly?

Comment (p0pr0ck5) [inline t/043-shdict.t]: Sure, was just copying the existing style from previous tests.

Also, i need to fix these tests to *_by_lua_block directives. Will do that within the next few hours.

File: t/043-shdict.t [hunk 2]
```diff
@@ -2448,3 +2448,216 @@ GET /test
...
+
+
+
+=== TEST 93: duplicate zones
+--- http_config
+    lua_shared_dict dogs 1m;
+    lua_shared_dict dogs 1m;
+--- config
+    location = /test {
+        content_by_lua '
```
Comment (membphis) [inline t/043-shdict.t]: Newly added test cases should use the *_by_lua_block form instead.

**Comment by p0pr0ck5** [comment]
No idea why there's a conflict here :|

**Comment by p0pr0ck5** [comment]
Ok, finally got a proper re-base done!

File: t/146-shdict-decr-init.t [hunk 3]
```diff
@@ -0,0 +1,248 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+use lib 'lib';
```
Comment (membphis) [inline t/146-shdict-decr-init.t]: this line is not necessary

File: t/146-shdict-decr-init.t [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    lua_shared_dict dogs 1m;
+--- config
+    location = /test {
+        content_by_lua_block {
+            local dogs = ngx.shared.dogs
+            for i = 1, 20 do
+                dogs:set("bar" .. i, i, 0.001)
+            end
+            dogs:set("foo", "32", 0.001)
+            ngx.location.capture("/sleep/0.002")
```
Comment (membphis) [inline t/146-shdict-decr-init.t]: How about use ngx.sleep directly?

File: t/146-shdict-decr-init.t [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    lua_shared_dict dogs 1m;
+--- config
+    location = /test {
+        content_by_lua_block {
+            local dogs = ngx.shared.dogs
+            for i = 1, 20 do
+                dogs:set("bar" .. i, i, 0.001)
+            end
+            dogs:set("foo", 32, 0.001)
+            ngx.location.capture("/sleep/0.002")
```
Comment (membphis) [inline t/146-shdict-decr-init.t]: DITTO

File: t/146-shdict-decr-init.t (hunk 4 above)
Comment (p0pr0ck5) [inline t/146-shdict-decr-init.t]: Yep, done. Thanks!

File: t/146-shdict-decr-init.t (hunk 3 above)
Comment (p0pr0ck5) [inline t/146-shdict-decr-init.t]: Whups, this was copied from test 144.There are a few new tests that have this, likely worth a separate PR to clean up these tests for style.

**Comment by p0pr0ck5** [comment]
The lua-resty-core relative for this PR is at https://github.com/openresty/lua-resty-core/pull/77

**Comment by agentzh** [comment]
@p0pr0ck5 I think we'd better only add this new `decr` method to lua-resty-core's FFI-based implementation. I do not want to maintain two parallel implementations for new API functions any more :)

**Comment by p0pr0ck5** [comment]
Are we then blocked on https://github.com/openresty/lua-nginx-module/issues/949? Or can we get this feature added in as-is, and then move forward with FFI-only implementations post 1.11.8.1? Personally I would like to see this go in now, but I'm biased ;)

**Comment by hongliang5316** [comment]
@p0pr0ck5 
Hello,
will you have a time to re-implement this feature via lua-resty-core's FFI method?

**Comment by p0pr0ck5** [comment]
Do you mean https://github.com/openresty/lua-resty-core/pull/77?

**Comment by hongliang5316** [comment]
@p0pr0ck5 
Hello,
Yeah, LGTM, thanks!

**Pull request closed without merging**